rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceSpec.scala

```diff
@@ -2059,4 +2056,284 @@ class ReduceSpec extends FlatSpec with Matchers with PersistentStoreTester {
     result.exprs should be(Seq(Expr(resultList)))
     errorLog.readAndClearErrorVector().unsafeRunSync should be(Vector.empty[InterpreterError])
   }
+  "toSet" should "transform [1, 2, 3] into Set(1, 2, 3)" in {
+
+    implicit val errorLog: ErrorLog[Task] = new ErrorLog[Task]()
+    val toSetCall: EMethod =
+      EMethod(
+        "toSet",
+        EListBody(
+          EList(
+            List[Par](
+              GInt(1L),
+              GInt(2L),
+              GInt(3L)
+            )
+          )
+        ),
+        List[Par]()
+      )
```

Comment:
We really should use the ParBuilder to build AST in this test.

Reply:
On one hand, using `ParBuilder` in tests for `Reduce` introduces unnecessary dependencies and violates separation of concerns - testing both the normalizer and the evaluator. On the another, using `ParBuilder` increases readability by reducing preparation. I side with SOC.

Reply:
Test setup is **not testing** the things used in test setup - it's setting up the stage for the test.

If a test fails in its setup - it's either due to a bug one has just introduced, or due to a change to a cross-cutting concern so drastic that the by-hand test setup would be obsolete and lead to the test no longer testing anything that can happen in production - making the test null in the worst case. It's like drawing a good conclusion from the wrong argument.

Heck, I bet some of the hand-crafted ASTs we have now in tests can't ever happen in production either (they're not sorted).

I don't think we understand separation of concerns the same way. If I need the input par to be normalized *and* sorted - does my calling `normalize` (which sorts underneath) violate SOC? (Hint: it doesn't.)

Hand-crafting complex input data for a test where using prod code in setup would suffice for the test's purpose is terribly misguided, detrimental to test reliability and test maintainability. I can't stress enough how wrong this is.

Reply:
> unnecessary dependencies 
quite the contrary, the fact that this test doesn't use ParBuilder for creating Reducer inputs (which is the only way a reducer input is produced in prod BTW) can lead to false positives in tests (par order can affect execution results / pattern matching; in general: ex falso quodlibet, or GIGO if you will).

Reply:
I side with Joe here. The ReduceSpec should test only the Reducer and the input of the reducer is the normalised AST, not the Rholang code.

If the unit tests are unreadable and unmaintainable then it's a problem in the API of the unit under test and the tests just point to that issue. Using tools to make the tests readable only hides the actual problem.

Reply:
> The ReduceSpec **should test only** the Reducer and the input of the reducer is the normalised AST, not the Rholang code.

How can I stress more that **test setup is not testing**? Is that a controversial claim?

> the normalised AST, not the Rholang code.

My claim is, normalization is so complex, that the only **reliable** way to have normalized test input is to use the normalizer. Similarly for parsing. Otherwise we end up doing a lot of test fixing when a detail not relevant to them changes.

Have you read my argument for why pretending that we know how to create the complex test input data is wrong until one cannot write the test otherwise?

---

Consider SRP - each piece of code should only have one reason to change. A test can change for the following reasons:
 1. input data needs to change
 2. the way we call the tested code changes
 3. output data needs to change

When we compare 'by-hand inputs' with 'prod-code inputs', the latter reduces the number of times tests need to change -> we have better, more SRP-y tests.

---

Consider Separation of Concerns - Dijkstra mentions it in the context of `thinking intelligently`, which is to limit the scope of consideration so that each aspect is considered in isolation; at any time we occupy ourselves only with one of the aspects. 

Using 'by-hand inputs' forces test reader & writer to think about how a normalized AST with a peek looks like. Using 'prod-code inputs' allows one to focus on the semantics of peek execution. The rholang source code, **is**, in an ultimate sense, the input to the reducer - and the only form of that input a human being can reason about *intelligently* when we consider results of execution.

So, 'prod-code inputs' leads to better, more SoC-y tests.

---

Granted, there are cases where input data needs to be constructed by hand, like e.g.:
 - making sure the reducer terminates as expected for invalid rholang ASTs (not a requirement, but an example)
 - making sure that the sorter outputs a sorted AST when given an not-sorted one (one could still use a test data generator, although without a *guarantee* that the generated data would be not-sorted)

Testing how a peek reduces though should not depend on how exactly the input AST looks like, as long as it's a normalized AST for a peek. More often than not, hand-writen AST will not be normalized. Further, a change to eek normalization could happen that changes execution results (be it a bug or a purposeful feature) - and the test for peek will not notice until it uses prod normalizer.

Reply:
if the test code is difficult to write then we have bad smell. Using tools to hide that smell is not helping at all. Instead we should refactor and eliminate the smell.

If we choose to have Rholang code for tests input then we are  not testing the reducer but instead we are testing the whole interpreter (parser/normalizer/reducer) so that test should not live in ReduceSpec but in some InterpreterSpec.

Now, thinking more about the issue, I would go in the opposite direction and  refactor each method (toSet, toMap etc) and move them outside the reducer and then unit-test them separately by calling the method apply directly, not by running the reducer.

Then the reducer will only contain one test per method, which would test that the actual method implementation is plugged in correctly.

Reply:
How is what I'm proposing difficult to write?

`checkResult("@0!([1,2,3].toSet())", "@0!(Set(1, 2, 3))")`

It's even easier than writing all of that in Rholang/RhoSpec.
It's also helping in at least 4 ways: readability, ease of writing, reliability (always uses what's **actually** normalized), and maintainability (less reasons for the test to change / SRP).

Maybe using prod code in setup is not using a tool? For sure it's using the ultimate thing that has a say what the test input should be: the code that produces that input in production.

---

> not testing the reducer but instead we are testing the whole interpreter (parser/normalizer/reducer) so that test should not live in ReduceSpec but in some InterpreterSpec.

I disagree that using prod code in test setup is a test for that prod code too. It does increase the fidelity level of that test though, and prevents testing impossible scenarios - and most tests aren't about those.

Moving the tests to InterpreterSpec would be :100: fine for me though, and I like it more than the other solution you proposed, for various reasons (incl. effort needed, test fidelity). Can we do that and close the discussion?

Reply:
These tests are doing exactly the same as the other methods tests are doing. So let's keep them like this for now and then we see how to refactor all the ReduceSpec tests in a separate PR.

Reply:
When will this PR come?
What's the outcome of this discussion?
What approach will this PR take?

I'd hate to restart this from scratch, so to me that's a no-no until we reach consensus on the above 3 questions.

To me, implementing the new tests in the/a InterpreterSpec would be an improvement over what's in the PR currently.

Reply:
As mentioned in the stand up, I'm OK with this PR proceeding as it is now. We should have the above discussion settled though before we write the next test that utilizes rholang AST without a strict necessity.